prompt2model/param_selector/search_with_optuna.py

```diff
@@ -0,0 +1,188 @@
+                    hyperparameters.get("evaluation_strategy", ["epoch", "no"]),
+                ),
+                "per_device_train_batch_size": trial.suggest_categorical(
+                    "per_device_train_batch_size",
+                    hyperparameters.get("per_device_train_batch_size", [4, 8, 16, 32]),
+                ),
+                "weight_decay": trial.suggest_loguniform(
+                    "weight_decay",
+                    hyperparameters.get("min_weight_decay", 1e-5),
+                    hyperparameters.get("max_weight_decay", 1e-1),
+                ),
+                "learning_rate": trial.suggest_loguniform(
+                    "learning_rate",
+                    hyperparameters.get("min_learning_rate", 1e-5),
+                    hyperparameters.get("max_learning_rate", 1e-2),
+                ),
+            }
+
+        # prepare the training args
+        # we are assuming here that the user will also provide these args with the additional
+        # args for range. Or we can provide an another argument of search_args (dict) that will
+        # tell the user to provide the arguments for doing hyperparamerter search
+
+        training_args = transformers.Seq2SeqTrainingArguments(
+            output_dir=hyperparameters.get(
+                "output_dir", DEFAULT_HYPERPARAMETERS.get("output_dir")
+            ),
+            logging_steps=hyperparameters.get(
+                "logging_steps", DEFAULT_HYPERPARAMETERS.get("logging_steps")
+            ),
+            save_strategy=hyperparameters.get(
+                "save_strategy", DEFAULT_HYPERPARAMETERS.get("save_strategy")
+            ),
+            num_train_epochs=hyperparameters.get(
+                "num_train_epochs", DEFAULT_HYPERPARAMETERS.get("num_train_epochs")
+            ),
+            per_device_train_batch_size=hyperparameters.get(
+                "per_device_train_batch_size",
+                DEFAULT_HYPERPARAMETERS.get("per_device_train_batch_size"),
+            ),
+            warmup_steps=hyperparameters.get(
+                "warmup_steps", DEFAULT_HYPERPARAMETERS.get("warmup_steps")
+            ),
+            weight_decay=hyperparameters.get(
+                "weight_decay", DEFAULT_HYPERPARAMETERS.get("weight_decay")
+            ),
+            logging_dir=hyperparameters.get(
+                "logging_dir", DEFAULT_HYPERPARAMETERS.get("logging_dir")
+            ),
+            learning_rate=hyperparameters.get(
+                "learning_rate", DEFAULT_HYPERPARAMETERS.get("learning_rate")
+            ),
+            predict_with_generate=True,
+        )
+        # training args here
+        trainer = self.trainer.trainer
+        trainer.train_dataset = training_sets
+        trainer.eval_dataset = validation
+        trainer.args = training_args
+
+        best_run = trainer.hyperparameter_search(
+            n_trials=5,  # FIXME: Discussion needed, where to put this arg and visibility for the user
```

Comment:
I think this should be set in the init function for this class.

Reply:
In the init function we only have the `trainer` param, so can I then change the definition?